app/src/main/java/org/openobservatory/ooniprobe/utils/NotificationsRouter.java

```diff
@@ -0,0 +1,70 @@
+package org.openobservatory.ooniprobe.utils;
+
+import com.google.firebase.messaging.FirebaseMessagingService;
+import com.google.firebase.messaging.RemoteMessage;
+
+import android.content.DialogInterface;
+import android.content.Intent;
+import android.util.Log;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.openobservatory.ooniprobe.R;
+import org.openobservatory.ooniprobe.activity.BrowserActivity;
+import org.openobservatory.ooniprobe.activity.MainActivity;
+import org.openobservatory.ooniprobe.activity.NotificationDialog;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Map;
+
+public class NotificationsRouter extends FirebaseMessagingService {
+
+    @Override
+    public void onMessageReceived(RemoteMessage remoteMessage) {
+        // TODO(developer): Handle FCM messages here.
+        // Not getting messages here? See why this may be: https://goo.gl/39bRNJ
+        Log.d(TAG, "Message from: " + remoteMessage.getFrom());
+        Log.d(TAG, "Message data payload: " + remoteMessage.getData());
+        Map<String, String> params = remoteMessage.getData();
+
+        // Check if message contains a data payload.
+        if (remoteMessage.getData().size() > 0) {
+            try {
+                //TODO we can use click_action instead of type
+                JSONObject data = new JSONObject(params.toString());
+                if (data.getString("type").equals("open_href")){
+                    Intent intent =  new Intent(getApplicationContext(), NotificationDialog.class);
+                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+                    intent.putExtra("message", remoteMessage.getNotification().getBody());
+                    intent.putExtra("payload", data.getString("payload"));
+                    getApplicationContext().startActivity(intent);
+                }
+            }
+            catch (Exception e) {
+                System.out.println("JSONException "+ e);
+            }
+        }
+
+        // Check if message contains a notification payload.
+        if (remoteMessage.getNotification() != null) {
+            Log.d(TAG, "Message Notification Body: " + remoteMessage.getNotification().getBody());
+        }
+
+        // Also if you intend on generating your own notifications as a result of a received FCM
+        // message, here is where that should be initiated.
+        /*
+        IntentRouter.getInstance(getApplicationContext())
+            .emit_string("orchestrate/notification",
+                remoteMessage.getNotification().getBody());
+         */
```

Comment:
Ah, so we're not using the `orchestrate/notification` thing? If so, we should perhaps remove the code, as it adds a number of lines of code that we can save, thus further simplifying the diff :-) - @lorenzoPrimi